lib/ukdebug/gdbstub.c

```diff
@@ -1106,3 +1113,78 @@ int gdb_dbg_trap(int errnr, struct __regs *regs)
 
 	return r;
 }
+
+static int gdb_setup_backing_console(void)
+{
+	/* Ignore breakpoints if the stub is not enabled on the command line. */
+	if (gdb_backing_console_id == GDB_BACKING_CONSOLE_NONE) {
+		uk_pr_info("Backing console not set\n");
+		return -ENXIO;
+	}
+
+	/* Ignore breakpoints if the backing console is not registered.
+	 * Maybe the backing console will be initialized later.
+	 */
+	if (gdb_backing_console_id >= uk_console_get_device_count()) {
+		uk_pr_info("Backing console not available\n");
+		return -ENXIO;
+	}
+
+	gdb_backing_console = uk_console_get(gdb_backing_console_id);
+	UK_ASSERT(gdb_backing_console);
+
+	if (gdb_virt_console_on()) {
+		uk_console_register(&gdb_virt_console);
+	} else if (gdb_virt_console_auto()) {
+		/* Must check before diabling the flags (see below) */
+		if (gdb_backing_console->flags)
+			uk_console_register(&gdb_virt_console);
+	}
+
+	/* If we didn't disable the STDOUT and STDIN flags, calls to
+	 * `uk_console_[out|in]` that happen outside of the gdb stub
+	 * could mess up the gdb communications. E.g., a call to
+	 * `uk_console_out` could write some bytes to the device that
+	 * look like garbage to the gdb host on the other end.
+	 * We assume here that the backing device is registered with
+	 * `ukconsole` exactly  once. This means we can also assume that
+	 * disabling the STDOUT and STDIN flags makes sure that the
+	 * underlying device is never used except explicitly stated with
+	 * `uk_console_[out|in]_with(gdb_backing_console, ...)`.
+	 */
+	gdb_backing_console->flags &= ~UK_CONSOLE_FLAG_STDOUT;
+	gdb_backing_console->flags &= ~UK_CONSOLE_FLAG_STDIN;
+
+	return 0;
+}
+
+void uk_debug_gdb_early_init(void)
+{
+	/* This is the first time that the stub is entered. */
+	UK_ASSERT(!gdb_backing_console);
+
+	uk_pr_info("Waiting for debugger (early) ...\n");
+
+	if (gdb_setup_backing_console() < 0) {
+		uk_pr_info("Skipping early breakpoint\n");
+		return;
+	}
+
+	gdb_arch_brk();
+}
+
+static int gdb_late_breakpoint(struct uk_init_ctx *ctx __unused)
+{
+	uk_pr_info("Waiting for debugger (late) ...\n");
+
+	if (gdb_backing_console)
+		uk_pr_info("Skipping late breakpoint: console already set up\n");
+	else if (gdb_setup_backing_console() < 0)
+		uk_pr_info("Skipping late breakpoint: failed to set up console\n");
+	else
+		gdb_arch_brk();
+
+	return 0;
+}
+
+uk_plat_initcall_prio(gdb_late_breakpoint, 0, UK_PRIO_LATEST);
```

Comment:
We could think of registering `gdb_late_breakpoint` to multiple inittab classes. This would enable early waiting n breakpoint for console drivers that are available earlier.

I think the breakpoint function should only once enter into a breakpoint.

Reply:
Again, I would say that we don't want to go back to multiple stages of gdb stub initialization. This is what we just moved away from by removing the early init mechanism. `ukprint` will introduce a print buffer, right? Then output won't even get lost. My opposition to your comment here is similar to this: https://github.com/unikraft/unikraft/pull/1479/commits/ffd403eb2cdb09c5ef0181a41214d6f307fa1fe3#r1758271414

Reply:
As further discussed in https://github.com/unikraft/unikraft/pull/1479#discussion_r1758983716, nothing to change for now. All good :+1: